File: 2020/js/day04.js
const read = require('./read.js')
const fn = read('04')

function passports() {
	let c = 0
	let i = 0
	let passport = []
	let res = []
	while (i < fn.length) {
		const str = fn[i]
		if (str == '') {
			res.push(passport)
			passport = []
		} else {
			passport = passport.concat(str.split(' '))
		}
		i++
	}
	
	return res
}

function part1(pps) {
	return pps.map(p => {
		if (p.length < 7) return false
		if (p.length == 8) return true
		const hasCid = p.map(v => v.substring(0, 3)).some(d => d == 'cid')
		return !hasCid
	}).filter(v => !!v).length
}

const keys = [
	'byr', 'iyr', 'eyr', 'hgt',
	'hcl', 'ecl', 'pid', 'cid'
]
const valid = p => {
	if (p.length < 7) return false
	const m = {}
	p.forEach(v => {
		const [key, value] = v.split(':')
		m[key] = value
	})
	let tmp
	const inValid = keys.map(key => {
		const value = m[key]
		switch(key) {
			case 'byr':
				if (!value) return false
				if (value.length != 4) return false
				tmp = parseInt(value)
				if (tmp < 1920 || tmp > 2002) return false
				break
			case 'iyr':
				if (!value) return false
				if (value.length != 4) return false
				tmp = parseInt(value)
				if (tmp < 2010 || tmp > 2020) return false
				break
			case 'eyr':
				if (!value) return false
				if (value.length != 4) return false
				tmp = parseInt(value)
				if (tmp < 2020 || tmp > 2030) return false
				break
			case 'hgt':
				if (!value) return false
				tmp = value.slice(-2)
				if (['in', 'cm'].indexOf(tmp) < 0) return false
				if (tmp == 'in') {
					tmp = parseInt(value.slice(0, -2))
					if (tmp < 59 || tmp > 76) return false
				} else {
					tmp = parseInt(value.slice(0, -2))
					if (tmp < 150 || tmp > 193) return false
				}
				break
			case 'hcl':
				if (!value) return false
				if (value.length != 7) return false
				if (value[0] != '#') return false
				break
			case 'ecl':
				return ['amb', 'blu', 'brn', 
					'gry', 'grn', 'hzl', 'oth'
				].indexOf(value) >= 0
				break
			case 'pid':
				if (!value) return false
				if (value.length != 9) return false
				break
			case 'cid':
				break
		}
		return true
	}).some(v => !v)
	return !inValid
}

function part2(pps) {
	return pps.map(valid).filter(v => !!v).length
}

console.log(part1(passports()))
console.log(part2(passports()))

File: 2020/hs/day04.hs
import Data.List (intersect)
import Data.List.Split (splitOn)

-- [["eyr", "2027"], ["hgt", "60in"]]
type Passport = [[String]]

pps :: [String] -> [Passport]
pps lines = keyvalue <$> group <$> groups
	where
		groups = splitOn [""] lines
		group g = concat $ fmap (splitOn " ") g
		keyvalue g = fmap (splitOn ":") g
		
requiredKeys = ["byr","iyr","eyr","hgt","hcl","ecl","pid"]
eclOptions = ["amb","blu","brn","gry","grn","hzl","oth"]

part1 :: [Passport] -> Int
part1 pps = length $ filter valid pps
	where
		keys g = fmap head g
		haveKeys g = (keys g) `intersect` requiredKeys
		valid g = length (haveKeys g) >= 7

within :: Int -> Int -> String -> Bool
within l h s= n >= l && n <= h
	where n = read s :: Int

lastN :: Int -> String -> String
lastN n xs = drop (length xs - n) xs

takeN :: Int -> String -> String
takeN n xs = take (length xs - n) xs

fieldValid :: [String] -> Bool
fieldValid ["byr", v] = within 1920 2002 v
fieldValid ["iyr", v] = within 2010 2020 v
fieldValid ["eyr", v] = within 2020 2030 v
fieldValid ["hgt", v]
	| unit == "cm" 	= within 150 193 num
	| unit == "in"  = within 59 76 num
	| otherwise = False
	where
		unit = lastN 2 v
		num = takeN 2 v
fieldValid ["hcl", '#':xs] 	= length xs == 6
fieldValid ["hcl", _] 			= False
fieldValid ["ecl", v] = v `elem` eclOptions
fieldValid ["pid", v] = length v == 9
fieldValid ["cid", _] = True
fieldValid [_, _] = False

part2 :: [Passport] -> Int
part2 pps = length $ filter joined pps
	where
		joined g = valid g && valid2 g
		valid2 g = all id $ fieldValid <$> g
		keys g = fmap head g
		haveKeys g = (keys g) `intersect` requiredKeys
		valid g = length (haveKeys g) >= 7

main :: IO ()
main = do
	fn <- lines <$> readFile "../res/04.input"
	let s = pps fn
	print $ part1 s
	print $ part2 s